components/raftstore/src/store/peer.rs

```diff
@@ -1610,22 +1607,31 @@ where
                 "disk_usage" => ?msg.get_disk_usage(),
             );
 
-            for index in msg
+            for (term, index) in msg
                 .get_message()
                 .get_entries()
                 .iter()
-                .map(|e| e.get_index())
+                .map(|e| (e.get_term(), e.get_index()))
             {
-                while let Some((propose_idx, instant)) = self.pending_propose_instants.front() {
-                    if index == *propose_idx {
+                if let Ok(idx) = self
+                    .proposals
+                    .queue
+                    .binary_search_by_key(&index, |p: &Proposal<_>| p.index)
```

Comment:
Do we need to measure the new overhead introduced here since this is a crucial path?
Besides, do we need to record the duration used in the `RaftClient` as the message is just put into the buffer here?

Reply:
The overhead of the binary search here should be okay.. `find_trackers` (which was find_propose_time` before) also uses binary search in critical paths already. Anyway, I'll measure the overall performance affect of the whole PR.

> Besides, do we need to record the duration used in the RaftClient as the message is just put into the buffer here?

It would be good to have. The time between `send` and `flush` is missing and counted as part of `commit_log_duration`. But I don't have a good idea to record it in `RaftClient`. This may change too much of the interfaces.